### PR TITLE
docs: forms pattern final draft

### LINKED ISSUE #329: Form pattern
Pattern describes the form element, it's multiple variants and its application for product and system level tasks. It covers the various inputs, actions and hierarchies that comprise a form and their usage, along with behaviors like errors and validation, progressive disclosure and inline verification. 

Questions will arise around how we separate the Form component and Form pattern in the same way we separate Shell components from Navigation patterns. The form and and of itself is more of a pattern comprised of components, not a component and possibly should not be categorized as such. It's often a self-contained UI, in which layotus and relationshiops are critical to user experience —  this aspect presents its own unique issues.

Carbon:
https://www.carbondesignsystem.com/components/form/code/

Cloud data and AI
https://github.ibm.com/CloudIntegrationDesign/DesignSystemAdoptionGuild/issues/38

Cloud data and AI
https://github.ibm.com/CloudIntegrationDesign/DesignSystemAdoptionGuild/issues/60

Cloud data and AI
https://github.ibm.com/CloudIntegrationDesign/DesignSystemAdoptionGuild/issues/62

IoT
https://pages.github.ibm.com/Watson-IoT/IoT-Design-Site/patterns/forms/

WCE
http://billboard1.fyre.ibm.com/component-library

- [x] Look at any existing form guidance from other BU's
- [x] Do any other components have relevant form guidance in their usage tab? (button does)
- [x] Create an outline for a Form pattern page.
- [ ] Create any supporting images.
- [x] When to use a number input vs. a select (when there's a large amount of options)

### PR DESCRIPTION
**Pull request by jeanservaas**
## docs: forms pattern final draft

Closes #329.

### CODE REVIEW

**Comment by joshblack** [comment]
Any idea why the deploy is broken? 😬

**Comment by jeanservaas** [comment]
@joshblack It's some issue with Now. A lot of the deploys randomly start working. cc @alisonjoseph  (Connor wrote this)

File: src/pages/patterns/forms-pattern/index.mdx [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+
+- [Notifications](/patterns/notification-pattern) <br/>
+- [Dialog](/patterns/dialog-pattern) <br/>
+
+</Column>
+</Row>
+
+## References
+
+- Joyce, Alita, [Tooltip Guidelines](https://www.nngroup.com/articles/tooltip-guidelines/), (Nielsen Norman Group, 2019)
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
- Alita Joyce, [Tooltip Guidelines](https://www.nngroup.com/articles/tooltip-guidelines/), (Nielsen Norman Group, 2019)
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+- Jakob Nielsen, [OK-Cancel or Cancel-OK? The Trouble With Buttons](https://www.nngroup.com/articles/ok-cancel-or-cancel-ok/), (Nielsen Norman Group, 2008)
+- Wroblewski, Luke, [Testing Accordion Forms](https://www.lukew.com/ff/entry.asp?1190), (A List Apart, 2010)
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
- Luke Wroblewski, [Testing Accordion Forms](https://www.lukew.com/ff/entry.asp?1190), (A List Apart, 2010)
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+- Whitenton, Kathryn, [Website Forms Usability: Top 10 Recommendations](https://www.nngroup.com/articles/web-form-design/), (Nielsen Norman Group, 2016)
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
- Kathryn Whitenton, [Website Forms Usability: Top 10 Recommendations](https://www.nngroup.com/articles/web-form-design/), (Nielsen Norman Group, 2016)
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+
+### Further reading
+
+- Coyle, Andrew, [Design Better Forms](https://uxdesign.cc/design-better-forms-96fadca0f49c), (UX Collective, 2016)
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
- Andrew Coyle, [Design Better Forms](https://uxdesign.cc/design-better-forms-96fadca0f49c), (UX Collective, 2016)
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+- Loranger, Hoa, [Form Design Quick Fix: Group Form Elements Effectively Using White Space](https://www.nngroup.com/articles/form-design-white-space/), Nielsen Normal Group, 2013)
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
- Hoa Loranger, [Form Design Quick Fix: Group Form Elements Effectively Using White Space](https://www.nngroup.com/articles/form-design-white-space/), Nielsen Normal Group, 2013)
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+- McCloskey, Marieke, [Accordions Are Not Always the Answer for Complex Content on Desktops](https://www.nngroup.com/articles/accordions-complex-content/), Nielsen Normal Group, 2014)
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
- Marieke McCloskey, [Accordions Are Not Always the Answer for Complex Content on Desktops](https://www.nngroup.com/articles/accordions-complex-content/), Nielsen Normal Group, 2014)
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+
+- [Text input](/components/text-input/usage) <br/>
+- [Password input](http://react.carbondesignsystem.com/?path=/story/textinput--toggle-password-visibility) <br/>
+- [Text area](http://react.carbondesignsystem.com/?path=/story/textarea--default) <br/>
+- [Checkbox](/components/checkbox/usage) <br/>
+- [Radio button](/components/radio-button/usage) <br/>
+- [Toggle](<(/components/toggle/usage)>) <br/>
+- [Button](<(/components/button/usage)>) <br/>
+- [Multiselect](http://react.carbondesignsystem.com/?path=/story/multiselect--default) <br/>
+- [Combobox](http://react.carbondesignsystem.com/?path=/story/combobox--default) <br/>
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
- [Combo box](http://react.carbondesignsystem.com/?path=/story/combobox--default) <br/>
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+| [Checkbox](components/checkbox/usage)                                                | To select or deselect one or more choices                                | Agree to terms and conditions, add optional items, select all that apply |
+| [Radio button](components/radio-button/usage)                                        | To select only one option from two or more choices                       | Pick type, shipping method, etc.                                         |
+| [Toggle](components/toggle/usage)                                                    | To choose one of two or more binary options                              | Changing user settings; On/off; Show/hide                                |
+| [File uploader](components/file-uploader/usage)                                      | To upload/attach a file or multiple files to a form                      | Attaching SSl certificates; adding config files to support tickets       |
+| [Combobox](http://react.carbondesignsystem.com/?path=/story/combobox--default)       | To select a single item (with typeahead functionality from a longer list | Choosing a state, country, or language preference                        |
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
| [Combo box](http://react.carbondesignsystem.com/?path=/story/combobox--default)       | To select a single item (with typeahead functionality from a longer list | Choosing a state, country, or language preference                        |
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+Second, the dedicated page form is not a modal and does not prevent the user from accessing their previous workflow. A back button will be available to them as part of the breadcrumb at the top of the page, or via the progress indicator component (if your form is part of a multi-step flow). The browser back button is also available. In short, back should never be an action on a secondary button. The secondary button is usually reserved for cancelling the task.
+
+Third, and most important, the top-pinned buttons create a very awkward relationship with the content when the user finishes the form and is ready to submit. If we feel it’s necessary to pursue pinned actions in the future, we should look into a pinned footer or tray to contain button groups.
+
+<Row>
+<Column colMd={4} colLg={4}>
+
+<DoDontExample
+type="do"
+caption="Do arrange primary and secondary buttons at the bottom of the form">
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
caption="Do arrange primary and secondary buttons at the bottom of the form.">
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 10]
```diff
@@ -0,0 +1,655 @@
...
+
+![Do arrange primary and secondary buttons at the bottom of the form](./images/do_bottom_buttons.png)
+
+</DoDontExample>
+</Column>
+<Column colMd={4} colLg={4}>
+<DoDontExample
+type="dont"
+caption="Do not top align primary and secondary action buttons in your layouts">
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
caption="Do not top align primary and secondary action buttons in your layouts.">
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 11]
```diff
@@ -0,0 +1,655 @@
...
+#### Naming actions
+
+Abstract terms like “Submit” give the user the impression that the form is generic. Although brevity is key in buttons, try to tell the user specifically what action your button will perform.
+
+<Row>
+<Column colMd={4} colLg={4}>
+
+<DoDontExample
+type="do"
+caption="Do use task-specific language in your buttons">
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
caption="Do use task-specific language in your buttons.">
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 12]
```diff
@@ -0,0 +1,655 @@
...
+
+![Do use task-specific language in your buttons](./images/Content_do.png)
+
+</DoDontExample>
+</Column>
+<Column colMd={4} colLg={4}>
+<DoDontExample
+type="dont"
+caption="Do not use vague language to describe an action">
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
caption="Do not use vague language to describe an action.">
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 13]
```diff
@@ -0,0 +1,655 @@
...
+- [city][state/province] [zip code]
+
+Avoid overloading users with too much information, when a multi-step form may be a better choice.
+
+<Row>
+<Column colMd={4} colLg={4}>
+
+<DoDontExample 
+type="do"
+caption="Do consider multi-step forms when faced with a lot of inputs">
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
caption="Do consider multi-step forms when faced with a lot of inputs.">
```

File: src/pages/patterns/forms-pattern/index.mdx [hunk 14]
```diff
@@ -0,0 +1,655 @@
...
+![Do consider multi-step forms when faced with](images/do_multi_step_form.png)
+
+</DoDontExample>
+</Column>
+
+<Column colMd={4} colLg={4}>
+
+<DoDontExample
+type="dont"
+caption="Do not overload the user with too many input controls at once, especially in modals ">
```
Comment (janchild) [inline src/pages/patterns/forms-pattern/index.mdx]:
```suggestion
caption="Do not overload the user with too many input controls at once, especially in modals.">
```

**Pull request merged**